packages/back-end/src/util/scheduleRules.ts

```diff
@@ -1,46 +1,33 @@
 import { ScheduleRule } from "../../types/feature";
 
 export const getCurrentEnabledState = (
-  scheduleRules: ScheduleRule[],
-  currentDate: Date
+  scheduledRules: ScheduleRule[],
+  now: Date
 ) => {
-  // If a feature doesn't have any rules, don't filter it out
-  if (!scheduleRules.length) {
-    return true;
-  }
-
-  // Ensure the scheduleRules array is always sorted earliest to latest
-  const sortedRules = scheduleRules.sort((a, b) => {
-    let dateA;
-    let dateB;
-    if (a.timestamp === null) {
-      dateA = a.enableFeature ? new Date("01-01-1971") : new Date("01-01-2090"); //TODO: Find a better way to handle this
-    } else {
-      dateA = new Date(a.timestamp);
-    }
-
-    if (b.timestamp === null) {
-      dateB = b.enableFeature ? new Date("01-01-1971") : new Date("01-01-2090"); //TODO: Find a better way to handle this
-    } else {
-      dateB = new Date(b.timestamp);
-    }
+  const sorted = scheduledRules
+    // Remove nulls
+    .filter((s) => s.timestamp)
+    // Convert timestamps to date objects
+    .map((s) => ({
+      date: new Date(s.timestamp as string),
```

Comment:
This is fine, but to answer your question about how to make typescript understand that you're filtering out nulls, you can use type guards (also called type predicates) - https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

```ts
// Type after the filter is applied
type ScheduleRuleNoNulls = Omit<ScheduleRule, "timestamp"> & { timestamp: Date }

const sorted = scheduledRules
  // Remove nulls
  .filter((s): s is ScheduleRuleNoNulls => !!s.timestamp)
  // `s.timestamp` is now Date, not Date|null
  .map((s) => ({
    date: new Date(s.timestamp)
```

Using `as` instead like you have it works, but is basically telling Typescript to ignore the type error that it sees.  If you later remove that `filter`, using `as` will continue to compile just fine, but will throw a runtime error.  Using type predicates, if you remove that filter it will throw a type error at build time instead, which is easier to catch.

Again, not a big deal in this instance since it's very unlikely to break in the future.  This is just for future reference.

Reply:
I went ahead and updated to use the type guard just for familiarity so if I need to do this again in the future I have a better concept.